src/errors.rs

```diff
@@ -1,619 +1,669 @@
+                            provided\n\n\
+                            {}\n\n\
+                            For more information try {}",
+                           Format::Error("error:"),
+                           Format::Warning(arg.to_string()),
+                           Format::Warning(num_vals.to_string()),
+                           Format::Warning(curr_vals.to_string()),
+                           suffix,
+                           usage,
+                           Format::Good("--help")),
+            kind: ErrorKind::WrongNumberOfValues,
+            info: Some(vec![arg.name().to_owned()]),
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn unexpected_multiple_usage<'a, 'b, A, U>(arg: &A, usage: U) -> Self
+        where A: AnyArg<'a, 'b>,
+              U: Display
+    {
+        Error {
+            message: format!("{} The argument '{}' was provided more than once, but cannot \
+                            be used multiple times\n\n\
+                            {}\n\n\
+                            For more information try {}",
+                           Format::Error("error:"),
+                           Format::Warning(arg.to_string()),
+                           usage,
+                           Format::Good("--help")),
+            kind: ErrorKind::UnexpectedMultipleUsage,
+            info: Some(vec![arg.name().to_owned()]),
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn unknown_argument<A, U>(arg: A, did_you_mean: &str, usage: U) -> Self
+        where A: Into<String>,
+              U: Display
+    {
+        let a = arg.into();
+        Error {
+            message: format!("{} Found argument '{}' which wasn't expected, or isn't valid in this context{}\n\
+                            {}\n\n\
+                            For more information try {}",
+                           Format::Error("error:"),
+                           Format::Warning(&*a),
+                           if !did_you_mean.is_empty() {
+                               format!("{}\n", did_you_mean)
+                           } else {
+                               "\n".to_owned()
+                           },
+                           usage,
+                           Format::Good("--help")),
+            kind: ErrorKind::UnknownArgument,
+            info: Some(vec![a]),
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn argument_not_found<A>(arg: A) -> Self
+        where A: Into<String>,
+    {
+        let a = arg.into();
+        Error {
+            message: format!("{} The argument '{}' wasn't found", Format::Error("error:"), a.clone()),
+            kind: ErrorKind::ArgumentNotFound,
+            info: Some(vec![a]),
+        }
+    }
 }
 
-impl Error for ClapError {
+impl StdError for Error {
     fn description(&self) -> &str {
-        &*self.error
+        &*self.message
     }
 
-    fn cause(&self) -> Option<&Error> {
-        match self.error_type {
+    fn cause(&self) -> Option<&StdError> {
```

Comment:
`cause` does not need to be provided. I don't know if this is used for something...